### ISSUE TITLE
Internal error: EACCES exception uncaught, opinions on how to catch it?

### ISSUE THREAD

**Issue by hyphenrf** [ocaml/dune#5468]
`XDG_RUNTIME_DIR` is set to a custom directory (`/var/run/...`) that doesn't exist, and which has a parent dir that the user has no privs for.. but dune doesn't fallback or catch that exception.

invoking `dune build -w` results in the following internal error:
```
/-----------------------------------------------------------------------
| Internal error: Uncaught exception.
| Unix.Unix_error(Unix.EACCES, "mkdir", "/var/run/user/<username here>")
| Raised by primitive operation at Stdune__fpath.mkdir in file "otherlibs/stdune/fpath.ml", line 12, characters 4-24
| Called from Stdune__fpath.mkdir_p in file "otherlibs/stdune/fpath.ml", line 23, characters 8-24
| Called from Stdune__fpath.mkdir_p in file "otherlibs/stdune/fpath.ml", line 34, characters 12-33
| Called from Stdune__fpath.mkdir_p in file "otherlibs/stdune/fpath.ml", line 34, characters 12-33
| Called from Dune_rpc_impl__run.run.(fun) in file "src/dune_rpc_impl/run.ml", line 86, characters 23-60
| Called from Fiber.Scheduler.exec in file "src/fiber/fiber.ml", line 854, characters 10-13
\-----------------------------------------------------------------------
```
but the watch mode blocks with `success` still.
Invoking without `-w` flag doesn't result in the same problem.

A possible patch I could start working on is one which adds a new error case for lack of perms, but how would the exception best be handled anyway? Should we panic and break execution? should we silently fallback to a default runtime dir?

bg info:
- `dune --version`: `3.0.2`
- `ocamlc --version`: `4.13.1`
- `Alpine Linux 3.14` system.

**Comment by rgrinberg** [comment]
I think the best that we can do here is gracefully shut down the rpc server and warn the user that it must provide a writeable XDG_RUNTIME_DIR

**Comment by hyphenrf** [comment]
would you like me to work on that?

**Comment by hyphenrf** [comment]
Hey so I made this little thing (#5478), feel free to check it out whenever

**Comment by Alizter** [comment]
@lukstafi Which version of WSL are you using? IIRC 1 has some issues with unix system calls, but 2 usually is smooth.

**Comment by lukstafi** [comment]
Anecdotally, I got the `Uncaught RPC Error` again today, and then run `eval $(opam env)`, I still got `Uncaught RPC Error` on the very next `dune build` but not on subsequent runs.

**Comment by lukstafi** [comment]
What I currently observe is: I consistently get `Uncaught RPC Error` when I run dune in watch mode: `dune build -w`, but not with a single run: `dune build`. That confused my reporting before.

**Comment by Alizter** [comment]
@lukstafi Could you share the exact error? Is it the same as the original issue?

**Comment by lukstafi** [comment]
Yes, I'm not worrying, it's not a "supported" WSL Linux distribution, I installed it from a docker image. I confirmed that things work with a WSL Linux distro from the Microsoft store (i.e. available via the WSL CLI).

**Comment by rgrinberg** [comment]
We are now handling this better by shutting dune down in such cases.
